### ISSUE TITLE
[feat] Push Notifications

### ISSUE THREAD

**Issue by sgammon** [tauri-apps/tauri#11651]
### Describe the problem

Several platforms supported by Tauri - macOS, iOS, and Android, to name a few - have some kind of built-in push notifications system. For Apple systems, of course there is [APNS](https://developer.apple.com/documentation/usernotifications/registering-your-app-with-apns), and for Android and Firebase-enabled apps, there is [FCM](https://firebase.google.com/docs/cloud-messaging).

Tauri isn't yet able to support these APIs, as far as I can tell. On Apple platforms, the developer must [register a method with the `AppDelegate` and call a method on `NSApplication`](https://developer.apple.com/documentation/usernotifications/registering-your-app-with-apns), which isn't easy to do from Rust. iOS uses a similar flow, but with `UIApplication` instead.

Push Notifications are likely to be a common need for Tauri developers, especially on mobile platforms.

### Describe the solution you'd like

I think Tauri should add calls to the `UIApplication` and/or `NSApplication` to register for push notification support on behalf of the developer; or, Tauri should make available such calls so the developer can implement support themselves.

### Alternatives considered

1) I looked for plugins and found none
2) I looked for easy Rust bindings to register for APNS, and found none because these APIs use `{NS,UI}Application`
3) I looked into alternative push systems (3rd-party), but these aren't integrated well with underlying operating systems like Android and iOS, as compared to native push infrastructure

### General support

In most cases, "registering" for these push systems consists of a few relatively simple steps:

1) "Registering" or "requesting" a token from the application, using system APIs, potentially within the app entrypoint
2) "Receiving" the token via some callback or delegate mechanism (the token is typically just some raw bytes or a string)
3) Making the received token available to the developer
  - Usually the developer will want to send this token to a server
  - Callbacks or even just a method to retrieve the latest push token (if any) would work fine for callers, probably

### Apple Platforms

**macOS**

It looks like [`tao` already sets up an `AppDelegate`](https://github.com/tauri-apps/tao/blob/88f0b4190f77d215f13e88f826c6b5bf06cafe57/src/platform_impl/macos/app_delegate.rs#L103-L107), where we would need to call [`registerForRemoteNotifications`](https://developer.apple.com/documentation/appkit/nsapplication/2967172-registerforremotenotifications), and then receive the token at the delegate method [`application(_:didRegisterForRemoteNotificationsWithDeviceToken:)`](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428766-application).

**iOS**

Nearly identical to macOS, but with [`UIApplication.registerForRemoteNotifications()`](https://developer.apple.com/documentation/uikit/uiapplication/1623078-registerforremotenotifications) and [the equivalent delegate method](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622958-application).

### Android

Firebase uses a `Service` on Android to implement messaging within an app. This service is declared in the `AndroidManifest.xml` and then implemented within app code; the service:

- Requests a token
- Receives the token/receives updated tokens
- Receives pushed messages directed to the app

This would be implemented in Kotlin as part of an Android mobile plugin for Tauri. Luckily there are no changes anticipated for `tao` or `tauri` itself; Android push can happen through the plugin alone.

### Windows

Windows apparently has [Windows Notifications System (WNS)](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/notifications/push-notifications/push-quickstart), which is accessible via the [`windows`](https://microsoft.github.io/windows-docs-rs/doc/windows/?search=PushNotifications) crate. So far it appears to function similar to Android, in that a library can obtain a push channel without special code in the app entrypoint.

### Linux

I'm not aware of native push infrastructure for Linux. Other SDKs can be used on Linux which do not require instrumentation in the app entrypoint.

**Comment by osdiab** [comment]
Someone published this undocumented plugin, but no idea if it works or what, haven’t tried it and don’t know anything about how mobile platforms work really.

https://github.com/guillemcordoba/tauri-plugin-fcm-push-notifications

Hoping that this feature gets added - seems like table stakes for Tauri to be usable for serious mobile apps!

**Comment by osdiab** [comment]
Not sure the current status of this but @gordongrech claims to have made FCM work on Android and iOS, so sounds like it’s feasible with the plugin system.

https://github.com/tauri-apps/tauri/discussions/8423

**Comment by eatenpancreas** [comment]
I have managed to get FCM working on Android, attempted IOS but didn't get far for now due to time constraints

This person has worked on a FCM plugin before and it seems to be further along on IOS. Haven't tested it myself but it could be useful:

https://github.com/flapili/tauri-plugin-fcm

**Comment by sgammon** [comment]
@osdiab / @eatenpancreas 

The blocker right now is iOS, which is structured differently in Tauri's internals than macOS (Mac is much more straightforward to implement). It may indeed be possible to implement some platforms with only a plugin, but Mac and iOS probably need this deeper integration for the reasons outlined in the PR (FCM does not use the same APIs as you guys probably know).

I've raised the iOS questions on Tauri's contributor Discord, but so far haven't found any expertise. I think what I will do is split off the iOS changes et. al so they can be worked on in a separate PR.

This would mean that macOS and Android, at least, can be delivered here, and then iOS and potentially WNS can follow shortly thereafter, after some advice from whoever wrote the iOS code.

**Comment by sgammon** [comment]
Also, I'd be happy to collaborate with anyone who is interested. My main hopes right now are to get this PR tree to stable, and to make this entire changeset less invasive by eliminating any changes that can be dropped.

**Comment by jd-carroll** [comment]
Nice work on all of this. As someone who is still trying to grasp the Tauri architecture, it might be nice to have a very brief write-up of the different pieces you needed to touch to make this happen (more for educational purposes). While I think all of the PR's are above, there seemed to be a lot of PR's to your personal repo too so it was a little hard to follow.

But broadly, nice work!

**Comment by jd-carroll** [comment]
Separately and why I am really here... Since this change involves accessing the `AppDelegate`, I'd like to make a plug for broader access or hooks into the `AppDelegate`. 

For example, I am building a banking app and I would like to display a loading / splash screen when the app is re-opened. The documentation I've found says I should hook into the `WillEnterForeground` event.

To my knowledge, this is not currently exposed in Tauri. It would be great to be able to redirect to a loading screen and then redirect back when this event is triggered.

I'm not trying to suggest that this be included in these changes, but merely that as this change involves the `AppDelegate` some consideration to additional `AppDelegate` use cases would be appreciated! :heart:

**Comment by davehorner** [comment]
commenting to voice support and to watch this ticket.  two repos and PRs given above.  I will play around, if someone has any notes for my cycle. lmk.

**Comment by nicoburns** [comment]
@jd-carroll You should take a look at the latest winit changes for macOS. It uses an event handler registration technique which means that Winit no longer needs to own the `AppDelegate` and applications can register their own. I libraries can also use the same technique that Winit is now using to register their own handlers without needing the `AppDelegate`.

**Comment by sgammon** [comment]
@davehorner Thanks. I think we are still waiting on advice from maintainers. I'm ready to finish the PR, but I haven't been able to locate anyone from the Tauri team who can advise about iOS changes. Until then, this PR sits idle. I can easily rebase it once that advice is forthcoming (or, a decision is made to cut iOS from this PR).

**Comment by sgammon** [comment]
@nicoburns Sorry, what is Winit? Is there a reason this feature/`tao` change is no longer needed? That would radically simplify the attached PR

**Comment by nicoburns** [comment]
> [@nicoburns](https://github.com/nicoburns) Sorry, what is Winit? Is there a reason this feature/`tao` change is no longer needed? That would radically simplify the attached PR

Winit is what Tao forked from. If Tao synced this change from Winit or Tauri switched back to using Winit then I think it would mean that this functionality could be implemented outside of Tao/Winit. At least on Apple platforms. I'm unfamiliar with how this works on windows.

**Comment by lindsaykwardell** [comment]
Commenting to follow, I'm working on a mobile app that will require push notifications in Android and iOS.

**Comment by jd-carroll** [comment]
@sgammon - Did you ever hear anything more from the Tauri team? (maybe on discord?)

@nicoburns - Thanks for the information! I tried looking at the recent changes / changelog and couldn't quite pinpoint where / when this change was made. Anything you could point me to? (either documentation, blog, changelog, etc.)

**Comment by jd-carroll** [comment]
@sgammon - Apologies for @mention'ing you again, but... I think you might be interested in this repo:
https://github.com/inKibra/tauri-plugins/blob/fc145aa0463f73a1e0e6f9a86d8e365c263c7ed3/packages/tauri-plugin-notifications/ios/Sources/NotificationsPlugin.swift#L138

Seems like it is (at the very least) an example of **how to integrate with the AppDelegate**.

**Comment by sgammon** [comment]
@jd-carroll Thanks and no worries for the multiple mentions. I'll take a look. Yes, the Tauri team got back to me on some of the stuff I asked (in particular, whether PRs for multiple crates is seen as invasive -- it's not, it's fine, which is great).

I still have questions about iOS integration, but either I'll answer those with links like you sent, or I'll continue without iOS and include it in a followup PR.

**Comment by bradleat** [comment]
I have a version here for iOS:

https://github.com/inKibra/tauri-plugins

**Comment by sgammon** [comment]
@bradleat Would you be interested in combining our efforts? That's really helpful

**Comment by d0m1n1kr** [comment]
@sgammon We successfully use https://github.com/inKibra/tauri-plugins for iOS and also added basic support for android here: https://github.com/FreshX-GmbH/tauri-plugins/tree/main/packages/tauri-plugin-notifications . The workaround  to not patch TAO (overwrite the app delegate in the plugin in proxy all calls to the original delegate) is a little bit hacky, but it works.

**Comment by Maxou44** [comment]
I think this feature is one of the last requirements to replace Capacitor / Cordova in many apps 🔥

**Comment by hrynevychroman** [comment]
Has someone set this for FCM? It is only needed for iOS for now. Maybe some examples would be great 😊

**Comment by poVoq** [comment]
Wouldn't it be possible to use WebPush?

**Comment by FabianLars** [comment]
not supported on mobile and requires a fairly recent macos version

**Comment by poVoq** [comment]
https://webkit.org/blog/13878/web-push-for-web-apps-on-ios-and-ipados/

Also available on Android.

**Comment by FabianLars** [comment]
This is only available in browsers though, not the webviews - and with that i also have to correct my first message as it seems like this is true for macos as well
https://developer.mozilla.org/en-US/docs/Web/API/Push_API#browser_compatibility

**Comment by adenta** [comment]
It seems like the only missing piece is to have a background task that can “listen” for a notification, there is already a robust notification sending plugin built in

**Comment by adenta** [comment]
Here is a plugin I'm working on, open for review comments: https://github.com/adenta/tauri-plugin-push-notifications/pull/1

**Comment by wushuang5112** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，双喜已收到您的邮件，期待与您再次联系！

**Comment by bradleat** [comment]
FYI: I have a version of push notifications working on iOS.  We are using this in production.

My only focus here was just iOS notifications:

https://github.com/inKibra/tauri-plugins/tree/main/packages/tauri-plugin-notifications

it looks like @d0m1n1kr  added android support here:

https://github.com/FreshX-GmbH/tauri-plugins/tree/main/packages/tauri-plugin-notifications

**Comment by adenta** [comment]
Nice

My focus ha s been on Hotwire native at the moment

**Comment by sheaksadi** [comment]
@bradleat @d0m1n1kr  can you guys add a section for android in the readme. I am pretty new to tauri would be great to know how to get this working

**Comment by seanaye** [comment]
Is there a reason that [the android support](https://github.com/FreshX-GmbH/tauri-plugins/tree/main/packages/tauri-plugin-notifications/src) has not been upstreamed into [the original ios support](https://github.com/inKibra/tauri-plugins/tree/main/packages/tauri-plugin-notifications)?

Would you be receptive to a PR?

**Comment by simonbeirouti** [comment]
this is like a cluster of suggestions 😅

Is there a definite and working plugin for push notifications on iOS and android? 

So far, there's:

https://github.com/tauri-apps/plugins-workspace/pull/2066 <- PR waiting

https://github.com/inKibra/tauri-plugins/tree/main/packages/tauri-plugin-notifications <- iOS only

https://github.com/FreshX-GmbH/tauri-plugins/tree/main/packages/tauri-plugin-notifications <- fork of the above, iOS & Android

Has there been any improvements to https://tauri.app/plugin/notification/ 

I haven't been able to go through the code as I'm still researching the package to use

**Comment by sgammon** [comment]
Hey guys, sorry, I really tried to get this over the line, and I am still happy to help answer questions if someone else is stuck, but the job I was at where I was using Tauri is long over. I asked for support via the Tauri Discord and got the cold shoulder, and basically left this where it was.

Sorry! I really do think Tauri should support push notifications as a built-in feature. If I can help advocate to get this work into Tauri mainline @FabianLars please feel free to tag me.

**Comment by sgammon** [comment]
(fwiw, one of the reasons the job evaporated was that we took a bet on tauri and it didn't have the infra we needed for basic things like notifications. that was definitely specific to our app's functionality, of course, and may have happened anyway, but the core team ignoring features like this and OSS devs who contribute them leads to real and actual decisions where the framework just isn't used anymore. just putting this out there for anyone who might also be considering the path we took.)

**Comment by Eveeifyeve** [comment]
> Hey guys, sorry, I really tried to get this over the line, and I am still happy to help answer questions if someone else is stuck, but the job I was at where I was using Tauri is long over. I asked for support via the Tauri Discord and got the cold shoulder, and basically left this where it was.
> 
> Sorry! I really do think Tauri should support push notifications as a built-in feature. If I can help advocate to get this work into Tauri mainline [@FabianLars](https://github.com/FabianLars) please feel free to tag me.

I feel like there should be in https://github.com/tauri-apps/plugins-workspace/tree/v2/plugins/notification

**Comment by NeoHuncho** [comment]
@Choochmeque has built a[ plugin for Tauri to enable push notifications](https://github.com/Choochmeque/tauri-plugin-notifications), I haven't tried it out myself yet, but it looks very promising!

**Comment by p1gp1g** [comment]
> > Wouldn't it be possible to use WebPush?
> This is only available in browsers though, not the webviews - and with that i also have to correct my first message as it seems like this is true for macos as well https://developer.mozilla.org/en-US/docs/Web/API/Push_API#browser_compatibility

For your information, it's possible to use web push to push to Android applications. It may be particularly useful for decentralized applications as the authentication can be done using VAPID (RFC8292, based on asymmetric keys), and unlike the FCM API, it doesn't require a single secret. cf. https://unifiedpush.org/news/20250131_push_for_decentralized/

It would be great having a tauri plugin that supports FCM and UnifiedPush by implementing the browser web push API directly. It would work out of the box for any website using web push

By the way, UnifiedPush works on Linux applications too. KUnifiedPush is developed by the KDE team: https://invent.kde.org/libraries/kunifiedpush and can be used on other DE as well.

**Comment by bradleat** [comment]
i've merged the android code push notification code here:

https://github.com/inKibra/tauri-plugins/pull/5 thanks to @mnlphlp and @seanaye 

the docs can probably use a touch of love. we'd also be interested in supported a unifiedpush fallback, but honestly i don't have much time to plan this.
